kafka-ui-api/src/main/java/com/provectus/kafka/ui/cluster/util/JmxClusterUtil.java

```diff
@@ -0,0 +1,99 @@
+
+import javax.management.*;
+import javax.management.remote.JMXConnector;
+import java.io.IOException;
+import java.math.BigDecimal;
+import java.net.MalformedURLException;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+@Component
+@Slf4j
+public class JmxClusterUtil {
+
+    @Autowired
+    private KeyedObjectPool pool;
+
+    private static final String JMX_URL = "service:jmx:rmi:///jndi/rmi://";
+    private static final String JMX_SERVICE_TYPE = "jmxrmi";
+
+    public static final String BYTES_IN_PER_SEC = "BytesInPerSec";
+    public static final String BYTES_OUT_PER_SEC = "BytesOutPerSec";
+    private static final String BYTES_IN_PER_SEC_MBEAN_OBJECT_NAME = "kafka.server:type=BrokerTopicMetrics,name=" + BYTES_IN_PER_SEC;
+    private static final String BYTES_OUT_PER_SEC_MBEAN_OBJECT_NAME = "kafka.server:type=BrokerTopicMetrics,name=" + BYTES_OUT_PER_SEC;
+
+    private static final List<String> attrNames = Arrays.asList("OneMinuteRate", "FiveMinuteRate", "FifteenMinuteRate");
+
+    public Map<String, BigDecimal> getJmxTrafficMetrics(int jmxPort, String jmxHost, String metricName) {
+        String jmxUrl = JMX_URL + jmxHost + ":" + jmxPort + "/" + JMX_SERVICE_TYPE;
+        Map<String, BigDecimal> result = new HashMap<>();
+        JMXConnector srv = null;
+        try {
+            srv = (JMXConnector) pool.borrowObject(jmxUrl);
+            MBeanServerConnection msc = srv.getMBeanServerConnection();
+            ObjectName name = metricName.equals(BYTES_IN_PER_SEC) ? new ObjectName(BYTES_IN_PER_SEC_MBEAN_OBJECT_NAME) :
+                    new ObjectName(BYTES_OUT_PER_SEC_MBEAN_OBJECT_NAME);
+            for (String attrName : attrNames) {
+                result.put(attrName, BigDecimal.valueOf((Double) msc.getAttribute(name, attrName)));
+            }
+        } catch (MalformedURLException url) {
+            log.error("Cannot create JmxServiceUrl from {}", jmxUrl);
+            closeConnectionExceptionally(jmxUrl, srv);
+        } catch (IOException io) {
+            log.error("Cannot connect to KafkaJmxServer with url {}", jmxUrl);
+            closeConnectionExceptionally(jmxUrl, srv);
+        } catch (MBeanException | AttributeNotFoundException | InstanceNotFoundException | ReflectionException e) {
+            log.error("Cannot find attribute from");
+            log.error(e.getMessage());
+            closeConnectionExceptionally(jmxUrl, srv);
+        } catch (MalformedObjectNameException objectNameE) {
+            log.error("Cannot create objectName");
+            log.error(objectNameE.getMessage());
+            closeConnectionExceptionally(jmxUrl, srv);
+        } catch (Exception e) {
+            log.error("Error while retrieving connection {} from pool", jmxUrl);
+            closeConnectionExceptionally(jmxUrl, srv);
+        }
+        finally {
+            if (srv != null) {
+                try {
+                    pool.returnObject(jmxUrl, srv);
```

Comment:
In case of error you invalidate an object from pool inside `closeConnectionExceptionally ` and then try to return it back to the pool. Does that work properly? Looks a bit weird.

Reply:
You haven't fixed that. I've made a small test and it shows that returning object after invalidating leads to `NullPointerException` or `IllegalStateException` depending on pool size
```
public static void main(final String[] args) throws Exception {
	KeyedObjectPool<String, Integer> pool = new GenericKeyedObjectPool<>(new BaseKeyedPooledObjectFactory<String, Integer>() {
		int counter = 0;

		@Override
		public Integer create(String key) throws Exception {
			return counter++;
		}

		@Override
		public PooledObject<Integer> wrap(Integer value) {
			return new DefaultPooledObject<>(value);
		}
	});

	String KEY = "X";
	pool.addObject(KEY);

	Integer v = pool.borrowObject(KEY);
	pool.invalidateObject(KEY, v);
	pool.returnObject(KEY, v);
}
```